File: CrackingTheCodingInterview/Chapter2/RemoveDups.java
// my implementation time complexity: O(n), Space complexity O(n)
void removeDups(LinkedListNode node) {
  if (node == null) return ;
  HashMap<Integer, Boolean> dups = new HashMap<Integer, Boolean>();
  dups.put(node.data, true);
  while(node.next) {
    LinkedListNode current = node.next;
    LinkedListNode next = node.next.next;
    if (dups.get(current.data))
      node.next = next;
    else
      dups.put(node.data, true);
    node = node.next;
  }
}

void removeDups(LinkedListNode node) {
  if (node == null) return ;
  HashSet<Integer> set = new HashSet<Integer>();
  LinkedListNode previous = null;
  while(node != null){
    if (set.contains(node.data)) {
      previous.next = node.next;
    } else {
      set.add(node.data);
      previous = node;
    }
    node = node.next;
  }
}


// FOLLOW UP Question
void removeDupsWithoutBuffer(LinkedListNode node) {
  if (node == null) return ;

  while(node) {
    LinkedListNode current = node;
    while(current.next != null) {
      if (node.data == current.data)
        current.next = current.next.next;
      else
        current = current.next;
    }
    node = node.next;
  }
}

File: Python/Practices/linked_list/remove_dups.py
"""
Remove Dups: Write code to remove duplicates from an unsorted li nked list. FOLLOW UP
How would you solve this problem if a temporary buffer is not allowed?
"""

# O(n) both time and space
def remove_dups(head):
    return None if not(head)
    unodes = {head.data}
    prev,curr = head,head.next
    while curr:
        if curr.data not in unodes:
            unodes.add(curr.data)
            prev,curr = curr,curr.next
        else:
            curr = curr.next
            prev.next = curr
    return head

# without temp buffer
def remove_dups_follow_up(head):
    return None if not(head)
    current = head
    while current.next:
        prev,fast = current,current.next
        while fast:
            if fast.data == current.data:
                fast = fast.next
                prev.next = fast
            else:
                prev = fast
                fast = fast.next
        current = current.next
    return head
